zerver/lib/test_classes.py

```diff
@@ -1103,3 +1109,189 @@ def setUp(self) -> None:
+                                        result)
+        self.verify_user_email_changed(action_key, user_profile.realm)
+
+    def test_unauthorized_email_change_from_confirmation_link_logged_in(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        action_key = self.create_email_change_confirmation_key(user_profile, new_email)
+
+        do_set_realm_property(user_profile.realm, 'email_changes_disabled', True)
+
+        self.login_user(user_profile)
+        result = self.do_email_change(user_profile, action_key)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Email address changes are disabled in this organization.",
+                                result)
+
+    def test_email_change_delivery_email_public_logged_in(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+
+        do_set_realm_property(
+            user_profile.realm,
+            'email_address_visibility',
+            Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
+        )
+        action_key = self.create_email_change_confirmation_key(user_profile, new_email)
+
+        self.login_user(user_profile)
+        result = self.do_email_change(user_profile, action_key)
+        self.assert_in_success_response(["This confirms that the email address for your Zulip"],
+                                        result)
+
+        self.verify_user_email_changed(action_key, user_profile.realm,
+                                       email_address_visibility_public=True)
+
+    def test_email_change_login_through_different_account(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        old_email = user_profile.delivery_email
+
+        action_key = self.create_email_change_confirmation_key(user_profile, new_email)
+
+        result = self.do_email_change(user_profile, action_key,
+                                      different_account=True)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Incorrect login. Login with your old email to change your email address",
+                                result)
+
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(new_email, user_profile.realm)
+        get_user_by_delivery_email(old_email, user_profile.realm)
+
+    def test_email_change_invalid_key(self) -> None:
+        user_profile = self.example_user('hamlet')
+
+        result = self.do_email_change(user_profile, 'invalid_action_key')
+        self.assert_in_success_response(["Whoops. The confirmation link is malformed."],
+                                        result)
+
+    def test_email_change_expired_key(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+
+        date_sent = timezone_now() - datetime.timedelta(days=2)
```

Comment:
Where in the code is it determined that an email change key is expired? I see other types of keys have  constants such as `INVITATION_LINK_VALIDITY_DAYS` but i dont see it for email change :thinking:

Reply:
It has the default validity set by this in `default_settings.py` `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS = 1`.

>Where in the code is it determined that an email change key is expired? 

All sorts of checks(in general for all confirmation key) including this are done by `get_object_from_key`. we use it to get the email change object twice in this PR, once in `handle_action_key` and `start_email_change`(we do it in this only to validate)

Reply:
Okay, thanks for checking! I think it'd be better to use `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS + 1` here, as that's clearer, rather than a magical value of 2 that will make folks puzzle over the same question i asked above :grinning:

Reply:
Ohh, makes sense. Thanks, will change to use `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS + 1`

Reply:
Changed.